app/code/Magento/Catalog/Model/ResourceModel/Product/Indexer/Price/Query/BaseFinalPrice.php

```diff
@@ -217,11 +217,7 @@ public function getQuery(array $dimensions, string $productType, array $entityId
         $select->where("e.type_id = ?", $productType);
 
         if ($entityIds !== null) {
-            if (count($entityIds) > 1) {
-                $select->where(sprintf('e.entity_id BETWEEN %s AND %s', min($entityIds), max($entityIds)));
-            } else {
-                $select->where('e.entity_id = ?', $entityIds);
-            }
+            $select->where('e.entity_id IN (?)', $entityIds);
```

Comment:
I believe that from the performance point of view better to use two where condition, to allow MySQL use range optimization for primary index scan:
like
```suggestion
             $select->where(sprintf('e.entity_id BETWEEN %s AND %s', min($entityIds), max($entityIds)));
             $select->where('e.entity_id IN (?)', $entityIds);
```
and return case for exact one product

Reply:
@kandy wondering why MySQL cannot do such kind of optimization by itself...

Reply:
Hi, thanks for reviewing my pull request!
I made an EXPLAIN on the two different queries with MySQL 5.5 and I don't see any difference. I also saw that in all the other indexes/resource models this optimization with "BETWEEN" is not used.
So I'm not sure about having more complicated code that introduce no positive value...
You have also tested the performance fix you did or just assumed it should work? Maybe I'm doing something wrong!

Let me know, thanks 💯 

P.S.: If we want to keep this fix, I would re-add the if-else on the number of elements in $entityIds, so if we have only one element I could put e.entity_id = X and only if two or more elements I would add your code.

Reply:
@orlangur, no, because it not common case and  base on specific usage in Magento Application

Reply:
I see, we know something about underlying data - like, range is not too big.

Still some heuristics could be applied on MySQL engine side.

Reply:
@kandy `e.entity_id IN (?)` I have serious doubts MySQL is not optimizing it well in case of only one ID.

Reply:
Using `BETWEEN` is wrong. Here is a real example. A catalog of 319,077 products, a product with an older `entity_id ` changes, as well as newer products. You get the following WHERE clause: `(e.entity_id BETWEEN 332999 AND 532434);`

In a multi-site installation etc... this SELECT returns nearly 2 million rows

```
mysql> SELECT COUNT(*) FROM `catalog_product_entity` AS `e`\n CROSS JOIN `customer_group` 
[...]
WHERE (ta_status.value = 1) AND (e.type_id = 'simple') AND (e.entity_id BETWEEN 332999 AND 532434);

+----------+
| COUNT(*) |
+----------+
|  1967373 |
+----------+
1 row in set (3.60 sec)
```

Reply:
@orlangur - maybe because of the limit of query size:
https://stackoverflow.com/questions/4275640/mysql-in-condition-limit/4275704

It was hell to find the problem.

Reply:
Sure, there can be limits with SQL, MySQL. config parameters etc... but this is what code is there for, to do intelligent things even with possible limitations somewhere else...

Reply:
@osakoova I don't see `e.entity_id IN (?)` part in your query.